### ISSUE TITLE
Preview of OpenTextEdit Integration

### ISSUE THREAD

**Issue by JuBan1** [notepadqq/notepadqq#704]
![img](https://i.imgur.com/G5iUaaU.png)

It's not done but I want to post it anyways. It's in a good enough shape that I started using it as my default editor.

A short overview:
* CodeMirror is gone, and with it QWebEngine and  most of the async code we've had to build.
* I'm now using KSyntaxHighlighting instead of my own solution. As a result there are now modes for nearly 300 languages available.

*Most* of the usual editor functionality works. Apart from a good deal of bugs the biggest issues are:

* No code folding
* No multiple cursor selection
* No themes other than a default light and dark theme.
* Opening larger documents (couple thousands loc+) is pretty slow since the highlighter still blocks the main thread.

I listed bugs I found in the TODO.txt

The diffstat is pretty crazy

`1692 files changed, 163923 insertions(+), 143376 deletions(-)
`

https://github.com/JuBan1/notepadqq/tree/ote

**Comment by JuBan1** [comment]
@Teklad You can give this a whirl now if you want.

The amount of outstanding issues I know of is fairly manageable by now:

* No multiple cursor selection.
* Some issues with Find & Replace (e.g. doesn't match \n yet).
* In Preferences, Line Height can't be set anymore (I don't know if this is feasible anymore).

But there must be dozens of bugs lurking around.

**Comment by Teklad** [comment]
There were a few build errors when I first went to try this but it was mostly syntax/naming issues.  There was also an undefined function "setTabWidth".  After getting all this sorted I gave this a test run.

From the looks of it most of the base features we have typically used seem to be there.  You even somehow managed to fix the slow syntax highlighting (or maybe it was fixed in a recent upstream release of ksyntaxhighlighting).

The lack of themes is a bit of an issue but from what I've seen implementing new themes doesn't seem like it would be too hard of a task.  The language support is definitely there no matter how you look at it.

As it stands this is almost more usable than the current QWebEngine branch.  At minimum the performance seems to be leaps and bounds better.  The logistical issues regarding that branch are quite a hassle.

I'd personally prefer to move forward with focus being on this component in the long run.  It would take a little extra time to get it up to speed but I believe the gains could definitely outweigh the loss.

**Comment by JuBan1** [comment]
Yes, sorry about the build issues! I've been working on it from two different computers, one of which doesn't have Qt installed. So I was hacking away at it in Notepad and used the git branch to upload my changes to test them at home. I should use a private branch for that..

I pushed a few new commits to the public branch. This includes fixing the syntax errors as well as patching a few other holes I found.

The syntax highlighting is still "slow" (~3000 lines/s on my machine), but I changed it so the Highlighter gives back control to Qt's event loop after highlighting each line. That way the GUI doesn't freeze while the document is gradually being highlighted.

**Comment by JuBan1** [comment]
*Note: I rebased the ote branch so if you want to test, you'll probably have to do a force pull* 

Making good progress. The OTE branch now compiles fine into a snap in the default environment. So Qt 5.5 is a sufficient minimum requirement.

Nearly all features are now re-enabled (indentation detection and printing being the most recent additions) and countless bugs lie slain by the wayside.

So.. to address the elephant in the room: we're in a bit of a pickle right now. Most of our users are stuck on the older, buggy WebKit version because we can't bring WebEngine to them easily. Since this doesn't seem to change anytime soon (and WebEngine has its own fair share of issues) I suggest we transition to using OTE quickly once it's reached the necessary level of quality (which, I hope, will be done in a few days, plus necessary testing).

The one big ticket to reaching feature parity with the current version is [multiple cursor/text selection](https://notepadqq.com/s/videos/multiselect.webm). Because of its complexity I suggest we leave it to after the integration. I want to get this feature done, but moreover I want to get Nqq back to a stable, unified base.

What you say, @danieleds @Teklad ?

**Comment by danieleds** [comment]
@JuBan1 I like how it's shaping.

About replacing it now, I don't want to rush: some features I particularly care about are indeed multiple cursors and Math mode, and I don't think removing them would be good, even if for just a few releases.
Every big change inevitably comes with bugs that have to be fixed with time, so I'd like to let our codebase settle a little bit for some time... We've got quite a lot of features but we never stopped to concentrate on bugfixing (either WebEngine bugs or different kind of bugs).

Replacing the core editor with OTE *now* feels like starting over again... better wait some more time ;)

**Comment by Teklad** [comment]
In my opinion the hasty swap to QWebEngine as a way to fix outstanding QWebKit bugs was pushed in a fairly hasty manner.  Now we've got a version in master that will barely compile for most users.  Our issue tracker inflated quite a bit without even getting to release state.
 
QWebEngine itself has a mountain of bugs as it is (1759 reported, and its still fairly new).  I'm keen on losing the math feature for a couple of releases if it means fixing stability.  Multiple cursor selection I can see, but the math feature doesn't strike me as vital (and honestly wouldn't hurt to be made as a plugin instead later on).

The master branch as it is already has no ETA whatsoever as to when we can even release an update that's even remotely up to quality due to QWebEngine having quite a few oddities.  If we ever make it to release with it I foresee a new flood of bug reports due to a lot of upstream bugs between different Qt versions.

At the end of the day I would prefer stability over a temporary loss of a feature or two.  The fact of the matter is most of the features will already be there.  I don't really see it as starting over.  The editor component itself isn't from scratch and most of the features you'd want in an editor are already covered for us, while the rest aren't particularly difficult to implement, though possibly time consuming.

**Comment by JuBan1** [comment]
> some features I particularly care about are indeed multiple cursors and Math mode, and I don't think removing them would be good, even if for just a few releases.

I can understand that. Neither are features I use much, so I don't miss them. I hope it won't take long to implement them; mcs is quite a piece of work, although I know it's possible since Qt Creators has it too.  

> we never stopped to concentrate on bugfixing (either WebEngine bugs or different kind of bugs).

I'd like to mention that OTE by itself already fixes a ton of outstanding issues (#365, #415, #466, #497, #416, #576, #323, #658, #464, #653, #542, #690, #701, #405), and makes it easy to implement plenty of requests that would otherwise require us to mess with CM's code (#148, #306, #140, #324, #441, #444, #234, let's just assume I listed them all :).

Leaving that aside, plenty of users can't enjoy our bugfixing if we can't get snaps/Ubuntu 16.04 to work. We've got more people on 1.4.2 than on any other version put together.

> Math mode

Any idea on how to graciously save it? I've only given it a quick glance and it appears to me as if it functions within a browser.js environment. That would mean it can't just run in a a Qml/JS container but needs Webkit/Webengine.

Best would be if it could render to an arbitrary surface. From there on I can blit the contents into OTE. 

If push comes to shove I might be able to render it in an offscreen webengine window. But then it would have to be some kind of plugin or we'd lose the advantage of being able to ship without WebEngine again..

**Comment by danieleds** [comment]
@JuBan1 

About Math mode, someone suggests [this](http://wiki.qt.io/Qt_and_LaTeX_via_KLFBackend), but I have no idea about its dependencies.

In general, it would be awesome being able to render it off-screen and then use some "generic OTE api" to embed it within the correct line/columns inside the editor. We could use this for potentially a lot of things (off the top of my head, git information and stuff like that)

**Comment by JuBan1** [comment]
[Well, it's getting somewhere, slowly.](https://gfycat.com/ZanyGoldenIndianjackal).

**Comment by danieleds** [comment]
That's nice, what approach did you follow?

**Comment by JuBan1** [comment]
I found a [handy library](https://github.com/jkriege2/JKQtPlotter) for Qt that can render to pixmaps without external dependencies.

I tried to replicate the current behavior, but it turns out manipulating the height of individual lines is *really* difficult to do. So I settled for looking for free space beneath the formula line to insert the image. It's not as nice as before but it's the best trade-off between usability and ease-of-development.

**Comment by Teklad** [comment]
I think there's a function which gets the current text height and such... Should be really useful in this scenario.  I can't remember the exact name of the function though.

**Comment by JuBan1** [comment]
Getting line or block height isn't the issue, but setting it is. You can see [in the original Mathjax issue thread](https://github.com/notepadqq/notepadqq/pull/625) how the line height grows to accomodate the larger math formula.

At first I tried replicating that but QPlainTextEdit doesn't really support setting the height of individual lines/blocks. Infuriatingly, changing that wouldn't be much of an issue but that piece of code is buried deep inside Qt5's source so I'd have to copy/paste at least half a dozen Qt classes into OTE and maintain them myself. And that's not gonna happen.

Then I played around with inserting artificial line-breaks  (like word-wrapping does) but ultimately that led to no solution I was satisfied with.

So I ended up doing what you saw in that webm.

**Comment by JuBan1** [comment]
I'm not dead yet! And OTE isn't dead yet either. I've been putting more work into an API to project arbitrary labels into the editor. That can be used for Math labels and other shenanigans.

[Progress is happening](https://gfycat.com/DeficientUnimportantCoral)

**Comment by Teklad** [comment]
That hex color clickable text is just downright fancy.  I don't think I've seen that feature before on a default editor (and I've used a few). :p

**Comment by JuBan1** [comment]
Okay, the labels stuff is in a pretty good spot now and I've pushed a rebase+update to https://github.com/JuBan1/notepadqq/tree/ote. For testing purposes LaTex- and Colorlabels are always enabled at the moment. 

@Teklad @danieleds Would you mind testing this again? Perhaps even installing it instead of your current Nqq install for a while? It'd be great to get some input on whether your day-to-day activities are negatively impacted by bugs/changes in functionality/etc.

That is, unless you can't live with multiple-cursor selection. That's the one big ticket still missing.

**Comment by Teklad** [comment]
@JuBan1 how's the status on this?  Sorry it's been a long time since I've checked in I got consumed by some other projects.  I'm excited to see this finally implemented... our bug list is piling up due to the stagnant state of the project.

**Comment by JuBan1** [comment]
My last comment tells most of the story. I've got some rudimentary code for multi-cursor selection now but nothing presentable yet. It's a fairly tricky feature because of its tight coupling with existing code for cursor and event handling.

If you want to lend a hand, my nqq-ote branch is ready and waiting :)

**Comment by Teklad** [comment]
Good news!  I'll finally have some free time this weekend to help with this.  Feature requests and bug reports are piling up insanely high.  We need to put a lot of focus on this before the entire application explodes, lol.

**Comment by JuBan1** [comment]
![We're getting somewhere](https://user-images.githubusercontent.com/4319621/45833226-78dbf500-bd04-11e8-9c95-f1fa3b06dbf0.gif)

**Comment by Teklad** [comment]
Care to give me a quick rundown on what's left to implement?  I'll try to put in some work for this throughout the next week or so.  Feel free to include any bugs that might need fixing as well.

**Comment by JuBan1** [comment]
The biggest help would be to just try the branch and find bugs. I try to fix them as I find them but Nqq just has such much functionality that could bug out by some change or another.

Non-bugs left to consider:

* Few themes available and I think some of them have badly chosen colors (solarized).
* Can't turn on/off math rendering or the hex-color labels yet.
* A few functions inside MainWindow are still stubs (e.g. `on_actionSpace_to_TAB_All_triggered`)
* Probably more, but that's what I can think of right now.

The list of bugs changes all the time but here's what I've got on my todo-list at the moment:

```
== Miscellaneous ==
- Left-facing selections cause constant redraws
- Don't composite extra events on *every* repaint.
- Asterisk shows up as an available Language inside Nqq, but doesn't have a definition file. Why?

== Multi-Cursor Selection ==
- Make sure the read-only attribute is respected during mcs operations
- Selections of multiple cursors shouldn't be able to overlap.
- Implement TextEdit::inputmethodevent

== Find & Replace ==
- Find might behave strangely for special chars (\t, \n, ..).

== Labels ==
- LatexLabels should only show up in comments
- Labels sometimes don't get painted when the line it's anchored to is moved downwards.
```

**Comment by JuBan1** [comment]
@Teklad If you'd like you can give Extensions a try too. I don't know if the changes break anything with what support we have for extensions and I don't want to break @qcerasmus 's stuff :) But I can't really run extensions on my system. Just doesn't want to work.

**Comment by Teklad** [comment]
I have yet to get extensions to work myself.  The process of getting it going is rather convoluted, which is why we don't ship any extensions with NQQ in the first place, lol.

**Comment by danieleds** [comment]
@JuBan1 WebEngine has settled but not in the hands of the users... yet :)

The main issue I see for merging OTE is actually still the same I expressed months ago: my concern is that the editor component should be quickly "hackable".

What I mean with that is that, for example, it took me something like an afternoon to implement math mode. Doing the same here obviously required a bit more time. If I wanted to add the "bookmarks" feature (just as an example), I could go in the codemirror directory, import some js, and then write a couple interfaces in c++. Here, I'd have to code it myself from scratch.

That's of course natural... CM is a project with 600+ contributors, and OTE is brand new. We couldn't expect anything different. But this also means that CM has an alive and independent community that improves the editor *in parallel* to what we're doing. And they're doing quite a lot of work for us. I don't want to lose that... I see a lot of value in that, something that we couldn't possibly match.

That said, yes, I agree, having to deal with web engine sucks a lot for a lot of reasons. However:
 - It is the path that most are following (e.g. vscode, atom). I'm not saying that we should do it because they're doing (actually, I don't like their approach, as I think a native UI is way superior). What I'm saying is that there are quite good reasons to do it.
 - It's not that it completely sucks. I mean, it worked fine for all of our increasing user base since we left Scintilla.

So there are pros and cons to the switch. I talked about the cons, but the pros are evident: I like the fact that it is way more lightweight, it doesn't require a whole web browser to work, and it is architecturally simple. But I have to evaluate the balance between the two and the risk I fear is stagnation down the road caused by the fact that we won't have the resources to do what CM is doing in parallel. This could change if somehow we grow and get a lot more regular contributors, as OTE proved to be definitely valid.

**TL;DR:** OTE is way cooler than CodeMirror but I don't want to risk project stagnation because we won't have the resources to do what CM is doing in parallel for us.

**Comment by JuBan1** [comment]
Bummer. I've been developing OTE with the primary prospect of ultimately having it merged into Nqq. (Though I don't complain being able to use it privately).

Thinking back, was there really all that much that we gained from having a hackable editor? I'm looking at the [commit history of the CM folder](https://github.com/notepadqq/notepadqq/commits/master/src/editor) and in the past two years there's been "math mode", "Font Selection" and "Show tabs/spaces/line symbols". Other than that it's just bug fixes or workarounds for iffy behavior, and apart from math mode those features were easy to code in C++ too. Sure, having the potential to add practically anything made in JS is great, but we've made little use of it in the past. (That bookmarks feature was requested [two years ago](https://github.com/notepadqq/notepadqq/issues/307). So "easy to implement" doesn't translate into "will be implementeed" either.)

What I'm worried about is that the stagnation is already creeping up on us. The list of open issues becomes longer, nearly all PRs are authored by us (except for l10n), and if our [stats are to be trusted](https://notepadqq.com/api/stat/view.html) Nqq is on a slow downwards trend (although I'm not sure if the stats are at all reliable for that). IMO the reason is that we have had v1.4.8 for more than four months now, yet most of our users are still on 1.4.2 because we can't even deliver it to them properly. We racked up near 200 commits after v1.4.8 that are just sitting there being useful no nobody. I'm sure it's as frustrating to you as to me.

Can you think of anything to do that would make OTE more appealing. Obviously I can keep adding stuff to it (like bookmarks) but that doesn't do much to address the root of your concerns. And while I don't intend on jumping ship anytime soon I can't guarantee anything of course.

**Comment by danieleds** [comment]
> We racked up near 200 commits after v1.4.8 that are just sitting there being useful no nobody. I'm sure it's as frustrating to you as to me.

This long delay was actually planned after the big switch to WebEngine. Actually, v2.0 should be out as soon as I clear this list: https://github.com/notepadqq/notepadqq/milestone/3
(#405 is an upstream bug and won't be fixed by us, it's there just as a reminder). The plan is to release it asap and after that we're probably going to merge the cmake work from @Teklad 

> Can you think of anything to do that would make OTE more appealing? Obviously I can keep adding stuff to it (like bookmarks) but that doesn't do much to address the root of your concerns.

Appealing *in the short term*? Probably not... I know that it's still young, but it's not mature enough yet. I'm actually using it, but a couple of bugs* showed up in just the first few minutes... they can be fixed, of course, but this already gives us an approximation of the expected stability. Don't take this bad: the work itself is remarkable (and man if it's smooth!)

Appealing *in the long term*? Well, that would be maturity. Maturity not just evaluated on features and bugs fixed, but rather in the sense of a piece of software that is regularly maintained and proves itself valid also *outside* notepadqq. That would be it.

\* I'll write them in the next reply

**Comment by Teklad** [comment]
If we're gonna talk about maturity... QWebEngine has a massive slew of bugs because its so new... and huge.  Also, adding an actual feature to CodeMirror that isn't already made is a massive pain in the rear.  Let's not forget the fact that the actual performance on it using a normal PC with an average processor is garbage.  It fails to load large files, etc.

On my system it takes well over a second to launch NQQ w/ QWebEngine with no saved session.  This is double the old branch.  The entire experience feels very sub-par and sluggish.  This is probably less noticeable on a newer system but the fact of the matter is why should a text editor ever have anything outside of basic system requirements?

The performance of my editor matters to me just as much as how stable it is.  I'd much rather deal with a missing feature than a sloth.  I don't mean to be overly critical here but we basically won the bug lottery when the webengine branch went live.

In my personal opinion it's way to new and prone to far to many bugs to be usable in production.  At least with a custom widget solution we can actually fix the bugs instead of having to wait a short millennium for the QT devs to get around to it.

**Comment by Teklad** [comment]
With all that said, I'm not against releasing the QWebEngine branch to production at some point, but I absolutely am with @JuBan1 on using OTE as a replacement asap once the bugs are ironed out (it can be done, we have the source :P).

**Comment by Teklad** [comment]
Actually if I'm being perfectly honest with myself, most of my contributions stopped because I've waiting in anticipation to see OTE merged.  I've been waiting for months now.  I think it's killing me a little inside.  Call me a fanboy.

To be honest I don't really understand java script code that well so the whole easily hackable thing depends on your skill set I suppose.  I mean I can make things work in JS but it takes me a whole lot longer than it would take @danieleds or basically anyone else with mediocre knowledge of JS (I have very little).

The main pro I've had with CM up to this point had nothing to do with features (it's basically any other editor), but more to do with it being easily available and saving a little bit of work. That's honestly a poor way to think when the original goal of this project was to be on par with N++.In this very specific instance, I feel that we may have cut a few corners that we probably shouldn't have for the sake of faster deployment.

I actually enjoyed the scintilla based editor as well back when that was a thing.  It worked fairly well for the features it had and the performance was fairly nice as well.  The only downside to Scintilla was adding a feature to that was by leaps and bounds a **real pain** compared to CM, but that's mostly due to the fact it is just written very awkwardly... and the actual API for it is older than grandfather time. You gotta think scintilla has been around since before the inception of Notepad++.

**Comment by JuBan1** [comment]
> Some of the bugs I encountered:

Thanks for telling me :+1: All three are fixed now.

The main issue isn't bug fixing or the lonesome development, it's just bug finding. Almost no bug takes much effort to fix. That's why I've been asking for help with bug hunting. I fix everything I see when interacting with it, but everyone has their own habits. E.g. all three issues you mentioned are plainly obvious but they're all outside my set of habits so I haven't encountered them. (And to redeem myself a little, that crash was an ASSERT I put in to catch a scenario I wasn't sure could happen. It wouldn't have crashed in release mode :P)

> Maturity not just evaluated on features and bugs fixed, but rather in the sense of a piece of software that is regularly maintained and proves itself valid also outside notepadqq.

Welp, that's probably not going to happen. While I may keep improving OTE, proving its use outside Nqq would basically mean to either create a new whole text editor myself or fork Nqq into a separate project, neither of which is a particularly appealing idea. 

I feel the same about QWebEngine as @Teklad (thanks for the backup by the way :)). I'm not too hot about dealing with anything on the JS-side of things, and I avoid the WebEngine branch because of the visual issues (there's still a white flash when opening a new tab that's very annoying with a dark theme) and the overall sluggishness (my laptop's running a pre-Haswell i5 mobile chip and it's not happy about running multiple web browsers).

Well, I've made my case. I think the advantages outweigh the disadvantages by a rather wide margin, but you're the boss. Perhaps something will come up that allows me to present OTE again in the future.

PS: That leaves me between a rock and a hard place. Spending my time improving OTE will cut down what time I contribute to Nqq (like it did in the past months). Continuing work on Nqq would mean leaving OTE in the dust and with it any chance of having it merged in the future.

**Comment by Teklad** [comment]
Well I see OTE as a more permanent solution.  QWebEngine was a necessary upgrade for the moment, but in all honesty our code base has gotten quite a bit more complicated just to support something that was basically forced upon us during the swap (thanks async).  I absolutely understand needing to squash bugs first, but so long as the features and goals align with our project (let's face it we barely use many of CodeMirror's features anyways), then I think the swap should happen.

**Comment by danieleds** [comment]
> I avoid the WebEngine branch because of the visual issues (there's still a white flash when opening a new tab that's very annoying with a dark theme) and the overall sluggishness

I totally agree with the feeling because I feel it myself

> we basically won the bug lottery when the webengine branch went live

I think I've been misunderstood with the webengine merge move. The plan was to merge it to master *so that* we could fix more bugs by having a bit more people use it. The bugs were expected :) And afaik they're actually all fixed now, if you exclude those couple of graphics glitches.

Anyway,

I think by now you guys know me a bit and you know that, whenever I have an opinion about something, you can make me change my mind with facts.

Now, believe me when I say that I'd like to change opinion about this. It's not like I'm in love with WebEngine :)  
If we were a lot of guys, maybe with secure funding behind, something like OTE would be the perfect choice. But we're not, so my priority is to avoid any possible future risk. The Editor is the most important part of nqq. I mean, we never did such a big change since scintilla. Even the QWebEngine switch was nothing compared to this. So you can see how this is quite a huge choice that will determine the future of the project, and should not be taken lightly.

I say, let's see how users respond to webengine (as I said, I hope to release v2 very soon). It is totally possible that people will start a rebellion against us*, and in that moment we'll know that we'll have to think again about this with new facts and evidence at hand. Or maybe everything will go fine and we'll have a stable editor that yes, is a bit sluggish, but will improve with time while we concentrate our focus on new things.


\* we have a very loud user base, trust me on this. In fact, I even received a death threat because of the snap package. lol

**Comment by Teklad** [comment]
We could offer them both side by side and let the users decide instead of forcing one or the other upon them.  This is a much better way to get a valid opinion than to just have the single offering.  I also believe the OTE version should be advertised for testing to see which one they prefer.

**Comment by Teklad** [comment]
So pros and cons of both real quick

**QWebEngine**
* Pros
    * Readily available editor
    * Easy to extend if you know javascript
    * Upstream support from CodeMirror
    * It's a web browser so anything that works in a browser works in the editor
* Cons
    * File loading performance poor
    * Rendering performance poor
    * Upstream bugs we can't fix
    * Memory usage is extremely high
    * Asynchronous operations are turning our code into a pig sty.  This could be remedied by a better implementation though

**OTE**
Pros:
    * Fix our own bugs instead of waiting for upstream (better than supplying workarounds)
    * Implement more advanced features down the road
    * Better performance
    * Much better memory footprint when dealing with multiple files
    * No rendering issues when resizing
    * Better options for plugin support
    * Possibility to unofficially support older distributions natively w/o snap
    * It gets the thumbs up from all the current developers (I had to throw it in there)
Cons:
    * Again, we fix our own bugs (This is both a pro and con)
    * Implementing features could be slightly more difficult
    * Risk of stagnation (but that applies to any project regardless of libraries including CM/WebEngine)
    * No upstream support
    * Possibly slower release cycle

With all this, here is what would make me happy:
* Allow the ote branch to live here in the main repository so making contributions is easier.
* Add an option to the configure script which will checkout the OTE branch and build that in master
* Offer a snap package for users that want to try OTE so we can get input (or distribute both binaries)
* Give us time to mature it to a point you're comfortable with it
* Discuss merging to master again later on

This isn't too far off from what we did when implementing QWebEngine itself outside of making it more publicly available.

I want to work on NQQ... but more specifically I want to work on the OTE branch which is actually usable on my computer.  I don't think it would be fair after all the work @JuBan1 has put in for it to be hidden away in his fork with no real means for users to try it outside of compiling from source.

I'm against forking into a separate project as that would split up the current developers into whichever one both the users and developers found more appealing.  There's been instances where this has almost completely killed off one or the other. LibreOffice (the OpenOffice fork) comes to mind here.  Keeping the community together means more developer power in the long-term.

This doesn't mean that contributions to the master branch will stop.  Bug fixes and features that aren't insane to implement can still be done in master and either rebased or ported into OTE.  This will, however, allow a little more leniency on code variations without having to deal with constant merge conflicts during development.

**Comment by JuBan1** [comment]
Another option that shot through my head was to compile both branches into separate binaries, ship them both to users, and give them a sort of "Experimental" option to switch from the default WebEngine to OTE executable.

But while I would be happy about an option that isn't as depressing as having OTE confined to my own little fork, I don't know if what you @Teklad or me just said is actually workable. It would be a big step towards getting OTE into a few users' hands to get the needed bug testing done. But there are drawbacks to all this:

* We'd receive bug reports for both branches here in Github (and I assure you some users are going to forget to tell us which variant they use :))
* Packaging two versions probably creates a fair amount of extra work?
* Users will have the expectation that we'll continue to support the OTE version. And while that's my intention, it's the same issue that @danieleds had with OTE to begin with, just on a smaller scale.
* Adding new features would have to be done with both versions in mind. Most of the time that's not difficult but in some cases vastly different solutions might come up (like CM's MathMode switch vs. OTE's plugin implementation).


> we have a very loud user base, trust me on this. In fact, I even received a death threat because of the snap package. lol

Wow. Some people... If there's anything to send death threats over it's clearly the Flatpak package!

**Comment by Teklad** [comment]
Regardless of what we use, however, there's always a chance for stagnation.  This is just how software development goes.  I searched around for notepadqq at one point on reddit a while back and recall seeing one of the few comments was "So sad it's written in javascript".  I'm sure they were referring to the fact we're basically trying to pass off a web app as a desktop application.

If we're forced to stick strictly with QWebEngine, then in my personal opinion it's not ready for release just yet.  The performance regressions as far as startup, new tabs, sessions, and general file loading go should be considered bugs that need fixing.  I've tested simplebrowser countless times and it doesn't run into any of the listed issues, but the resize bug is still there(obviously).

As much as I hate to say it... I'm almost certain this has to do with doing asynchronous operations improperly yet again.  There's a very specific callback style of flow that it expects and when you work against the grain there it causes some serious hiccups from all my previous testing.  

For comparison I got the **old** webengine branch webengine-beta working again to test.  Outside of being a little buggy it has a startup time of under 400ms, which is actually faster than the webkit branch for me.  I'm not saying the way I was doing it was right (it wasn't), but I tried to stick with the actual flow of how it expected to be called.  This actually gave a performance gain in this case.

My biggest thing with the OTE editor is once all the base features are implemented it's done outside of bug fixing and the occasional addition.  The reason I say this is that with a good enough plugin API features can be developed for it regardless whether we're the ones writing them or not.  This is basically in line with any decent editor.

For plugins that meet quality standards we can actually merge them into master as an optional component.  Since it's not strictly a javascript based editor we're talking about here, there are many many ways we can go about supporting a good plugin architecture.

For instance the math mode is useful, but by no means should it be loaded by default or be anywhere in the core editor code.  This is one of those very peculiar features that are only useful to a specialized group of people.  Features of this type do not belong in the base code anywhere if we can help it.

**Comment by danieleds** [comment]
> For comparison I got the old webengine branch webengine-beta working again to test. Outside of being a little buggy it has a startup time of under 400ms, which is actually faster than the webkit branch for me. I'm not saying the way I was doing it was right (it wasn't), but I tried to stick with the actual flow of how it expected to be called. This actually gave a performance gain in this case.

Are you saying that there is a way to fix the slugginess of the current implementation?  
Some time ago I extracted our core editor (the WebEngine one) and put it into a standalone application by itself. Not very useful of course, but it was really fast. Together with your observation, it seems that the slowness is caused by something which is *not* structurally part of the editor itself.

Now, the current async implementation is very solid, trust me on this. In fact, if we wanted we could even rewrite the whole async stuff into a synchronous API in a qtwebkit-style. This means that the big suspect is the amount of messages exchanged between nqq and the editor. If I remember correctly, you were caching a lot of things in your implementation.

> For plugins that meet quality standards we can actually merge them into master as an optional component. Since it's not strictly a javascript based editor we're talking about here, there are many many ways we can go about supporting a good plugin architecture.

But remember... if we're going more and more cross-platform, having compiled plugins is not the best option! :)

**Comment by Teklad** [comment]
Who said anything about compiled plugins?  The plugin API itself doesn't need to require a compiled language.

**Comment by danieleds** [comment]
I may have missed something here, what do you mean exactly with "the plugin API"?

**Comment by Teklad** [comment]
Just expose the editor API (and maybe some NQQ functions) itself to something like python, JS, etc....  This is more or less how most programs support actual plugins but don't want the overhead of worrying about cross-compatibility between different OSes.  These are of course examples and by no means an end-all.

**Comment by Teklad** [comment]
For your other question, the standalone editor didn't have mainwindow and friends communicating with it constantly either.  The caching I implemented in the old webengine branch was to actually reduce the stupid amount of JS calls we were making by providing that information in real time sent directly from JS to CPP.

I removed an entire layer of indirection... as it stands we're doing this:

* ask editor JS for information
* cpp editor function returns qtpromise object which waits for a reply
* mainwindow waits for qtpromise object to be fulfilled

My implementation:

* Js sends to cache
* Mainwindow gets information from editor cache

Waiting is probably a poor term here, but somewhere in the code (I'm not sure where), we're obviously waiting on something.  It's really hard to spot the bug though since I'm not familiar at all with QtPromise and the library itself appears to be fairly new (May 2017).

I'm not blaming QtPromise itself, just pointing out that the slowness has been there ever since the switch to using it started.... we're simply doing something wrong and I can't for the life of me figure out what.

**Comment by Teklad** [comment]
Found the slow startup issue:
```cpp
    void Editor::setFocus()
    {
        m_webView->setFocus();
       // asyncSendMessageWithResultP("C_CMD_SET_FOCUS")
    }
```
commenting out that line makes startup WAY faster.

**Comment by danieleds** [comment]
Wow!

Did you notice if the startup time scales with the number of tabs restored from session?

**Comment by Teklad** [comment]
Loading time for multiple-tab sessions is still not that great.  The CPU/resource usage when opening multiple tabs simultaneously is too high, lol.  That portion is going to be a lot harder to fix without figuring out a way to make the editor work from a single QWebEngine instance.  It's just a LOT of threads are initialized with each instance.

**Comment by Teklad** [comment]
Are we using any wait calls in sessions right now?  The startup time goes back up to stupid levels with even one open file in a session.

**Comment by Teklad** [comment]
Yep we are... markDirty, markClean, etc. cause waits.  It boils down to we can't use synchronous operations **at all** when using QWebEngine.  The only instance where this is okay is while closing notepadqq.

**Comment by danieleds** [comment]
Yeah, about sessions... this is what I discovered.

DocEngine::loadDocuments(), which is used by the session-loading code, is all nicely async. But if you check, the whole method is called once per each file in the session.

**Comment by danieleds** [comment]
About this specific case... instead of calling markDirty / markClean as we go, we could store them and then apply all of them at the end, in parallel.

This is more or less what your cache was doing, but specific to this case.

**Comment by danieleds** [comment]
But this is really only editor's business and no-one else, so probably some sort of agnostic caching mechanism would actually be better.

**Comment by Teklad** [comment]
We could use a QVariantMap to store cached values from JS.  This is *slightly* slower than the method I used originally but keeps the code quite a bit cleaner.

**Comment by danieleds** [comment]
Yes, we could. I want to think a bit about this because I'm not sure that caching everything unconditionally would be safe... we should have some sort of "cachable properties" that are safe to use even if the value is outdated. I'm open to ideas

**Comment by danieleds** [comment]
A new issue is better though; I don't want to mess with this one as this is intended for OTE

**Comment by Teklad** [comment]
Feel free to write one up.  I got a pretty decent list in my head of what should be cached vs. what should not.

**Comment by brupelo** [comment]
Hi guys, really interesting thread this one! it's been a very nice reading. Haven't been able to test that awesome OTE fork as I'd need to spin up some linux box. One fast question, what's the status of multicursor feature in OTE? In fact, I'd like to know how it compares with sublimetext or codemirror. I'm not mentioning here qscintilla as the multicursor there has been poorly implemented.

Specially I'd like to know how ote multicursor deals with undo and redo... Are the selections preserved in the undo history? To me a proper multiselection feature would be one that behaves 1:1 as sublimetext or codemirror. The concepts there have been good ones from the very beginning... ;)

Also, I guess there isn't an easy way to test the ote fork on windows7 easily, right? 

Thanks in advance!

**Comment by JuBan1** [comment]
@brupelo Hey,

> One fast question, what's the status of multicursor feature in OTE?

It's done so far. It's similar but a little different to CodeMirror's. I can't compare it with Sublime since I don't use it. If OTE ever piques any interest (like being included into Nqq) the details might change a bit. But so far I'm happy with it and use it all the time. 

> Are the selections preserved in the undo history?

They are not, but selections aren't preserved for normal undos either. I like the idea though it should be done always or never.

> Also, I guess there isn't an easy way to test the ote fork on windows7 easily, right?

The OTE component itself shouldn't have any trouble running on Windows, but the rest of Notepadqq will revolt :)

**Comment by brupelo** [comment]
> It's done so far. It's similar but a little different to CodeMirror's. I can't compare it with Sublime since I don't use it. If OTE ever piques any interest (like being included into Nqq) the details might change a bit. But so far I'm happy with it and use it all the time.

When it comes to multiple selections SublimeText does the job perfectly... it was the first one implementing this [feature](https://forum.sublimetext.com/t/history-of-simultaneous-multi-caret-editing/43253) and it just did it correctly from the very beginning.

> They are not, but selections aren't preserved for normal undos either. I like the idea though it should be done always or never.

Well, not sure how difficult would be to have but not even Scintilla (who's been created since ancient ages) hasn't implemented multiselection the right way... in fact, the more you use multi-selection with Scintilla the more you realize the feature is more like a hack, and of course, it doesn't implement undo history with multiselection.

If undo history was implemented in OTE your widget would reach the "next level" in terms of quality and the editor would be a serious choice in Qt land to be a nice standalone text editor.

> The OTE component itself shouldn't have any trouble running on Windows, but the rest of Notepadqq will revolt :)

Well, if you ever create a mcve/simple example of usage with OTE where you can compile/run it on windows just let me know and I'll give it a shot... and not only that, if I found the component to be usable I wouldn't mind to create bindings to use either on pyqt or pyside2. There isn't many good choices we python coders can cosider as standalone text editor... and to be fair with you, there isn't any strong reasons to embed chromium using QtWebEngineView on your qt apps, not only because it's slow as hell but also because it's buggy, here's a little 5min tests embedding codemirror/ace [https://github.com/brupelo/mcves](https://github.com/brupelo/mcves), native apps ftw! :)

**Comment by maverick74** [comment]
Please forgive me for not understanding it, but... 
Are we sticking with KSyntaxHighlighting or with OpenTextEditor?

**Comment by JuBan1** [comment]
> Please forgive me for not understanding it, but...
> Are we sticking with KSyntaxHighlighting or with OpenTextEditor?

OpenTextEditor is built using a modified KSyntaxHighlighting. The current mainline Notepadqq uses CodeMirror, and it seems we'll stick with that one for the foreseeable future.

**Comment by JuBan1** [comment]
Yes, it's a big bummer and it's probably what's responsible for my lack of contributions over the last months. I use Nqq+OTE privately and any change basically means integrating it into both Nqq+OTE *and* Nqq+CodeMirror which only makes it harder to maintain OTE as a fork.

I do understand the reasons for not using it; it's a risk adopting a new and largely untested piece of software. What we disagreed on is whether the risk was worth taking.

**Comment by brupelo** [comment]
> I do understand the reasons for not using it; it's a risk adopting a new and largely untested piece of software. What we disagreed on is whether the risk was worth taking.

This is a again my personal opinion so take it with a bit grain of salt, if we were talking about company where external forces such as customers/clients/money/deadlines were messing up with software development i'd understand you wanted to go with low-risks choices and yeah, of course having bug-free stable deps with years of development would be the obvious choice. And yeah, if the product would have to offer technical support to thousands of customers you wouldn't want to make releases with big impact.

But... this is an open source where you guys are just having fun (if I've understood correctly) and don't want to create any commercial endeavour... So, I'm not sure I understand why you didn't started used the whole job of your team member... And sure, there will be bugs that with codemirror you wouldn't have, features you wouldn't have... But in exchange you'd have had a highly motivated team-member willing to keep up with code mirror (conclusions after reading the whole thread :)) , so IMHO the decission of sticking with codemirror doesn't look smart to me...

Also, I think with these tiny software projects having full control with the tech is key so you can deliver features much faster, by being attached with codemirror the workflow/pace should be (theorically) much slower...

But again, take this as a personal opinion, at the end of the day i didn't even test notepadqq so... :) . I just got curious with this thread cos I saw OTE and I considered there was a lot of potential of becoming a nice standalone widget... the fact it was highly coupled with linux drove me off though ;)

**Comment by maverick74** [comment]
@JuBan1 So, let me get this strait: to use KSyntaxHighlighting one has to use OpenTextEditor, because KSyntaxHighlighting is integrated into the later, right?

If so, i must say i'm all in for KSyntaxHighlighting. Specially now that Qt Creator is using it as well ( https://cullmann.io/posts/qt-creator-4-9-uses-ksyntaxhighlighting/ ), we'll probably see new features and performance improvements in it in near times.

**Comment by maverick74** [comment]
@brupelo Notepadqq is not just for Linux ( see bug #169 ) - it's just missing a small push so that we can have it on Windows as well :)

**Comment by danieleds** [comment]
Hopping in real quick. As @JuBan1 rightly says, my main concern has always been for the future of an Editor component which is maintained by just one person. I just want to clarify that I'm not a-priori against any change.

A more in-depth reply will follow.

**Comment by maverick74** [comment]
@danieleds ok... so, what about using KTextEditor? 

You could also outsourcing it, it is well maintained, uses KSyntaxHighlighting and it's qt based...

**Comment by brupelo** [comment]
> @brupelo Notepadqq is not just for Linux ( see bug #169 ) - it's just missing a small push so that we can have it on Windows as well :)

Nice! Thing is... if notepadqq became fully cross-platform (ie: windows not a 2nd citizen anymore) and codemirror was replaced by OTE (or any other native rock-solid native widget) I'd be sold to give it a shot straightaway to this project... truth is, nowadays there are no real options in qt land when it comes to have a proper hackable text editor... my concept of proper text editor is something with similar (or higher) quality to SublimeText, other than that  I'll call it suboptimals solutions :)

**Comment by JuBan1** [comment]
@maverick74 

Yes, the KSyntaxHighlighting framework is used for highlighting when using the OTE component. I only modified it a little to suit our needs. It'll also benefit from upstream patches and updates when I rebase it (thanks for the link by the way. I didn't know QtCreator now uses KSH).

That also takes care of a fair amount of the maintenance burden but not of all of it. There's still a good chunk of stuff that CodeMirror used to do which I replaced on my own. The [math renderer](https://github.com/notepadqq/notepadqq/pull/625) is a good example which took me days to replicate in OTE.

Regarding KTextEditor: I'm not sure but I think KTE depends on a lot of other KDE stuff. We probably don't want this. And on a more personal note, if I wanted to let KTextEditor do all the work I could just use Kate ;)

@brupelo I doubt Nqq will ever be as good as Sublime. We're a hundred miles away from having feature parity, with or without OTE.

**Comment by maverick74** [comment]
@JuBan1 Yes, using KTE would mean ending up having (_almost_) Kate with a new name. 
I mentioned it just for the case you also wanted to remove the burned of maintaining OTE or as a backup plan in case you get too much busy IRL.

_From what i understand_, you're right about KTE dependencies relaying a lot on KDE ( https://api.kde.org/frameworks/ktexteditor/html/ktexteditor-dependencies.html )... but maybe i'm getting it wrong...

All in all, Nqq+OTE+KSH would be a great improvement if you ask me :)

**Comment by danieleds** [comment]
> @danieleds ok... so, what about using KTextEditor?

Unfortunately, all the KDE dependencies as @JuBan1 said. Find me another modern editor component which is fast and in line with our mission and I'm on it. Unfortunately, there doesn't seem to be one.

I think I expressed my line of thought enough times in this thread. OTE is a project which totally deserves to be written, for the reason expressed in the previous paragraph. Unfortunately, I can't put it in the official Nqq build today, because OTE has high ambitions but it is still young.

If something has changed since the last time I tried it (i.e. most of the bugs are now fixed, the code is in a stable state, there are no regressions with respect to our last release, and new features can be provably implemented without too much effort) then great, let's create an official experimental Notepadqq release with OTE and get user feedback.

Are we there?
@maverick74 , @brupelo , are you already relying on OTE on a daily basis?

I see that in @JuBan1 's repo there is a very long list within "*Upstream issues that are fixed in OTE*". I honestly think that's great work from @JuBan1 (no doubt about it since he's been one of the best contributors in terms of quality and attention to details) but it can also be misleading, since most* of those issues have not been fixed "thanks to OTE" but just because @JuBan1 decided to work there instead of here (again, I have no problem with that, I just felt I needed to clarify for those not familiar with the code).

Sorry if I sound like a stubborn evil dictator, but difficult decisions need to be taken in order to avoid the risk of falling into chaos. "Spare-time open-source" is rarely long-lived.

----------------------
\* not all

**Comment by JuBan1** [comment]
I invite anyone to grab Nqq-OTE from my [github](https://github.com/JuBan1/notepadqq) (need to compile from source) and use it. It's practically up-to-date with Nqq-master (excepting the last changes to localization) and runs great for me. 

I fixed those bugs that I came across. But there are sure to be bug left that I didn't come across. Reporting bugs is the best thing to get OTE accepted, aside from bribing danieleds with loads of money.

**Comment by danieleds** [comment]
> I invite anyone to grab Nqq-OTE from my github (need to compile from source) and use it.

I, too, encourage anyone interested to use it and help polish it.

And also to send me all ur moneyz $$$

**Comment by brupelo** [comment]
> Sorry if I sound like a stubborn evil dictator, but difficult decisions need to be taken in order to avoid the risk of falling into chaos. "Spare-time open-source" is rarely long-lived.

@danieleds No, you're not, we all understand you only pursue the best for the project, all people who's interested on this small niche of custom text editors also wish that :) . My only point was... if there are nice contributors in your team, try to encourage them using their work or giving feedback... we coders like people using our software, don't we? So even if their solutions are still WIP... what's so wrong if you use their job? In fact, read what Juban1 said before:

> Yes, it's a big bummer and it's probably what's responsible for my lack of contributions over the last months. I use Nqq+OTE privately and any change basically means integrating it into both Nqq+OTE and Nqq+CodeMirror which only makes it harder to maintain OTE as a fork.

If I was in a similar position to him I'd got pissed off and probably created my own project out of OTE evolving independently of notepadqq... in any case, I think it wasn't a good choice not taking action few months ago, probably at this point OTE would have become more solid thanks to feedback/debugging from users.

> I invite anyone to grab Nqq-OTE from my github (need to compile from source) and use it. It's practically up-to-date with Nqq-master (excepting the last changes to localization) and runs great for me.

@JuBan1 Well, I'm a total geek about trying text editor standalone editors... if you ever make it easy to compile & run on window just ping me and I'll give it a shot... not only that, I could create also a pyqt/pyside2 wrapper of your component so testing&debugging would be done 100x times faster than in c++ ... (I'm a c++/python coder btw)

> Reporting bugs is the best thing to get OTE accepted, aside from bribing danieleds with loads of money.

@JuBan1 Well, as I said before, keeping history with multiselection is something both Sublime & Codemirror do... you said OTE isn't doing it so I'd call that a bug (we're striving to replace codemirror, right?) ... or if you don't want to call it a bug, let's then call it a "missing feature" then ;)

> And also to send me all ur moneyz $$$

You're creating open-source and only sticking to linux... that sounds fishy... lol :'D

**Comment by headkaze** [comment]
> I invite anyone to grab Nqq-OTE from my [github](https://github.com/JuBan1/notepadqq) (need to compile from source) and use it. It's practically up-to-date with Nqq-master (excepting the last changes to localization) and runs great for me.
I much prefer your version due to the performance gains. Are there any plans to move the official version to OTE?

**Comment by JuBan1** [comment]
> > I invite anyone to grab Nqq-OTE from my [github](https://github.com/JuBan1/notepadqq) (need to compile from source) and use it. It's practically up-to-date with Nqq-master (excepting the last changes to localization) and runs great for me.
> 
> I much prefer your version due to the performance gains. Are there any plans to move the official version to OTE?

Nope, sorry. Both versions have effectively seized development. I'm still using my variant personally but that's about it.

**Comment by brupelo** [comment]
> Nope, sorry. Both versions have effectively seized development. I'm still using my variant personally but that's about it.

@JuBan1 Out of curiosity, has OTE development being continued since last time? Couple of years ago I'd asked you whether OTE would preserve undo/redo selection like SublimeText/CodeMirror/Monarch et al but at that time you told:

> They are not, but selections aren't preserved for normal undos either. I like the idea though it should be done always or never.

I think this is pretty important core feature in case you want to allow users building commands on top of it (like sublime text) but probably difficult to implement, isn't? 

Anyway, your project was really cool and it's sad it hasn't been merged to this cool notepadqq project... In my experience QtWegEngineView is just too slow to use... I've made some fast tests using codemirror/ace + pyqt5/pyside2 and not good feeling :/

**Comment by JuBan1** [comment]
@brupelo 

Not really. It works for my use cases and so I haven't had much drive to work on it.

> I think this is pretty important core feature in case you want to allow users building commands on top of it (like sublime text) but probably difficult to implement, isn't?

I think it would have been non-trivial at least. The undo-redo functionality is provided by the TextEditor widget, so modifying it would probably have meant replacing it entirely.
